### PR TITLE
Enabling ILC reflection for certain types in System.Net.Primitives.

### LINKED ISSUE #20482: ILC optimization causes reflection to fail in System.Net.Primitives CookieParser
test `SetCookies_InvalidData_Throws` is throwing a Debug.Assert in Debug uapaot ilc test runs, so the test will be disabled in order for the test run to complete. This issue is to track the work of fixing the issue causing the Debug.Assert to be called, and for re-enabling the test.

### LINKED ISSUE #22056: Some of the HttpListenerResponseCookiesTests are failing with NullReferenceException in UAPAOT
System.Net.Tests.HttpListenerResponseCookiesTests.SetCookie_CookieDoesntExist_ThrowsArgumentException [FAIL]
        System.NullReferenceException : Object reference not set to an instance of an object.
        Stack Trace:
              at xunit.console.netcore!<BaseAddress>+0xb456d0
              at xunit.console.netcore!<BaseAddress>+0xb453ed
              at System.Net.Tests.HttpListenerResponseCookiesTests.<SetCookie_CookieDoesntExist_ThrowsArgumentException>d__11.MoveNext()
           --- End of stack trace from previous location where exception was thrown ---
              at xunit.console.netcore!<BaseAddress>+0x95a051
              at xunit.console.netcore!<BaseAddress>+0x959fee
              at xunit.console.netcore!<BaseAddress>+0x959f63
           --- End of stack trace from previous location where exception was thrown ---
              at xunit.console.netcore!<BaseAddress>+0x95a051
              at xunit.console.netcore!<BaseAddress>+0x959fee
              at xunit.console.netcore!<BaseAddress>+0x959f63
           --- End of stack trace from previous location where exception was thrown ---
              at xunit.console.netcore!<BaseAddress>+0x95a051
              at xunit.console.netcore!<BaseAddress>+0x959fee
              at xunit.console.netcore!<BaseAddress>+0x959f63

### PR DESCRIPTION
**Pull request by CIPop**
System.Net.HttpListener is temporarily (actual fix tracked by #13607) relying on reflection to access internal Cookie and CookieCollection members from System.Net.Primitives.
This is allowing the temporary workaround in UWPAOT.

/cc @Priya91 

Fixes #20482
Fixes #22056

### CODE REVIEW

**Comment by CIPop** [comment]
@dotnet-bot Test Outerloop Windows x86 Release Build
@dotnet-bot Test Outerloop Linux x64 Release Build
@dotnet-bot Test Outerloop UWP CoreCLR x64 Debug Build
@dotnet-bot test Outerloop UWP NETNative x86 Release Build

**Comment by morganbr** [comment]
@CIPop , did you hit exceptions that made you need rd.xml for this? If so, please pass them on to @shiftylogic , who's filling in some of the NS 2.0 reflection gaps.

File: src/System.Net.Primitives/src/Resources/System.Net.Primitives.rd.xml [hunk 1]
```diff
@@ -0,0 +1,22 @@
+<?xml version="1.0" encoding="utf-8" ?>
+<Directives xmlns="http://schemas.microsoft.com/netfx/2013/01/metadata">
+  <Library Name="*System.Net.Primitives*">
+    <Assembly Name="System.Net.Primitives">
+       <!-- VSO 449560 The following types are used by HttpListener. -->
+       <Type Name="System.Net.Cookie" Browse="Required Public" Dynamic="Required Public">
```
Comment (morganbr) [inline src/System.Net.Primitives/src/Resources/System.Net.Primitives.rd.xml]: Browse is redundant -- Dynamic covers it.

File: src/System.Net.Primitives/src/Resources/System.Net.Primitives.rd.xml [hunk 2]
```diff
@@ -0,0 +1,22 @@
...
+         <Method Name="InternalSetName" Dynamic="Required" />
+         <Field Name="IsQuotedDomain" Dynamic="Required" />
+         <Field Name="IsQuotedVersion" Dynamic="Required" />
+
+         <!-- Used by CookieExtensions -->
+         <Method Name="ToServerString" Dynamic="Required" />
+         <Method Name="Clone" Dynamic="Required" />
+         <Property Name="Variant" Dynamic="Required" />
+       </Type>
+       <Type Name="System.Net.CookieCollection" Browse="Required Public" Dynamic="Required Public">
```
Comment (morganbr) [inline src/System.Net.Primitives/src/Resources/System.Net.Primitives.rd.xml]: Same comment on Browse

File: src/System.Net.Primitives/src/Resources/System.Net.Primitives.rd.xml [hunk 3]
```diff
@@ -0,0 +1,22 @@
+<?xml version="1.0" encoding="utf-8" ?>
+<Directives xmlns="http://schemas.microsoft.com/netfx/2013/01/metadata">
+  <Library Name="*System.Net.Primitives*">
+    <Assembly Name="System.Net.Primitives">
+       <!-- VSO 449560 The following types are used by HttpListener. -->
+       <Type Name="System.Net.Cookie" Browse="Required Public" Dynamic="Required Public">
+         <!-- Used by CookieParser -->
+         <Method Name="InternalSetName" Dynamic="Required" />
```
Comment (morganbr) [inline src/System.Net.Primitives/src/Resources/System.Net.Primitives.rd.xml]: The Dynamic="Required Public" line above should mean you don't need extra directives for the members.

**Comment by CIPop** [comment]
> @CIPop , did you hit exceptions that made you need rd.xml for this?

The tests were failing initially and this was the plan discussed with @davidsh . 
Are you saying an rd.xml is not needed if we make these public? 
Can we leave them as-is (internal) and use the rd.xml only?

**Comment by morganbr** [comment]
TL;DR, ignore my question and go ahead with this change.

@CIPop, to both of your questions: not exactly.

Making these public just removes the block on reflectability. It's still up to ILC whether to actually add metadata or not. I was just curious whether it was doing it automatically or if it needed hints from rd.xml for the patterns you wrote. I suspect you needed the rd.xml because you used NS 2.0 reflection APIs that @shiftylogic is working on adding ILC smarts for.

File: src/System.Net.Primitives/src/System/Net/Cookie.cs [hunk 4]
```diff
@@ -65,8 +65,34 @@ public sealed class Cookie
         private int m_version = 0; // Do not rename (binary serialization)
 
         private string m_domainKey = string.Empty; // Do not rename (binary serialization)
-        internal bool IsQuotedVersion = false;
-        internal bool IsQuotedDomain = false;
+
+/* 
+    TODO: #13607
+    VSO 449560
+    Reflecting on internal method wont work on AOT without rd.xml and DisableReflection
```
Comment (DavidGoll) [inline src/System.Net.Primitives/src/System/Net/Cookie.cs]: >wont [](start = 34, length = 4)

nit: won't (same for other instances throughout)

Comment (CIPop) [inline src/System.Net.Primitives/src/System/Net/Cookie.cs]: Good catch! Forgot to install the spellchecker after installing VS2017...

File: src/System.Net.Primitives/src/System/Net/Cookie.cs [hunk 5]
```diff
@@ -65,8 +65,34 @@ public sealed class Cookie
...
+    block in toolchain.Networking team will be working on exposing methods from S.Net.Primitive
+    public,this is a temporary workaround till that happens.
```
Comment (DavidGoll) [inline src/System.Net.Primitives/src/System/Net/Cookie.cs]: >,this [](start = 10, length = 5)

nit: missing space (same for other instances)

**Comment by CIPop** [comment]
Outerloop UWP CoreCLR x64 Debug Build had an unrelated System.Diagnostics.Process failure.

File: src/System.Net.Primitives/src/Resources/System.Net.Primitives.rd.xml (hunk 1 above)
Comment (CIPop) [inline src/System.Net.Primitives/src/Resources/System.Net.Primitives.rd.xml]: [offline conversation] - I've removed the Type directives completely: my intention was to only expose the methods/properties, not the entire type.

File: src/System.Net.Primitives/src/Resources/System.Net.Primitives.rd.xml (hunk 3 above)
Comment (CIPop) [inline src/System.Net.Primitives/src/Resources/System.Net.Primitives.rd.xml]: Fixed

File: src/System.Net.Primitives/src/Resources/System.Net.Primitives.rd.xml (hunk 2 above)
Comment (CIPop) [inline src/System.Net.Primitives/src/Resources/System.Net.Primitives.rd.xml]: Fixed

**Comment by CIPop** [comment]
@dotnet-bot test Outerloop UWP NETNative x86 Release Build
